File: quant1x/exchange/code.rs
use crate::exchange::SecurityType;
use std::fmt;

// Note: Exchange codes are exposed via `ExchangeCode` constants below.

// ExchangeCode newtype mirrors Go's `ExchangeCode` (string alias) and provides
// conversion to `ExchangeId`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ExchangeCode(pub &'static str);

impl ExchangeCode {
    pub const fn new(s: &'static str) -> Self {
        ExchangeCode(s)
    }

    pub fn as_str(&self) -> &str {
        self.0
    }

    pub fn id(&self) -> ExchangeId {
        match self.0 {
            x if x == EXCHANGE_SZSE.as_str() => ExchangeId::ShenZhen,
            x if x == EXCHANGE_SSE.as_str() => ExchangeId::ShangHai,
            x if x == EXCHANGE_BSE.as_str() => ExchangeId::BeiJing,
            x if x == EXCHANGE_HKEX.as_str() => ExchangeId::HongKong,
            x if x == EXCHANGE_US.as_str() => ExchangeId::USA,
            _ => ExchangeId::Unknown,
        }
    }
}

impl std::fmt::Display for ExchangeCode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

// Common ExchangeCode constants (match Go names)
pub const EXCHANGE_UNKNOWN: ExchangeCode = ExchangeCode::new("unknown");
pub const EXCHANGE_SSE: ExchangeCode = ExchangeCode::new("sh");
pub const EXCHANGE_SZSE: ExchangeCode = ExchangeCode::new("sz");
pub const EXCHANGE_BSE: ExchangeCode = ExchangeCode::new("bj");
pub const EXCHANGE_HKEX: ExchangeCode = ExchangeCode::new("hk");
pub const EXCHANGE_US: ExchangeCode = ExchangeCode::new("us");

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExchangeId {
    Unknown = 255,
    ShenZhen = 0,
    ShangHai = 1,
    BeiJing = 2,
    HongKong = 21,
    USA = 22,
}

impl fmt::Display for ExchangeId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            ExchangeId::Unknown => "unknown",
            ExchangeId::ShenZhen => EXCHANGE_SZSE.as_str(),
            ExchangeId::ShangHai => EXCHANGE_SSE.as_str(),
            ExchangeId::BeiJing => EXCHANGE_BSE.as_str(),
            ExchangeId::HongKong => EXCHANGE_HKEX.as_str(),
            ExchangeId::USA => EXCHANGE_US.as_str(),
        };
        write!(f, "{}", s)
    }
}

#[derive(Debug, Clone)]
pub struct ExchangeInfo {
    pub id: ExchangeId,
    pub code: String,
    pub name: String,
    pub description: Option<String>,
    pub is_active: bool,
}

impl ExchangeInfo {
    pub fn new(code: &str, name: &str, desc: Option<&str>, id: ExchangeId) -> Self {
        Self {
            id,
            code: code.to_string(),
            name: name.to_string(),
            description: desc.map(|s| s.to_string()),
            is_active: true,
        }
    }

    pub fn validate(&self) -> Result<(), String> {
        if self.code.is_empty() {
            return Err("exchange code cannot be empty".into());
        }
        if self.name.is_empty() {
            return Err("exchange name cannot be empty".into());
        }
        Ok(())
    }
}

impl fmt::Display for ExchangeInfo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}({})", self.name, self.code)
    }
}

#[derive(Debug, Clone)]
pub struct SecurityCode {
    pub market: ExchangeId,
    pub symbol: String,
    pub typ: SecurityType,
}

impl SecurityCode {
    pub fn new(market: ExchangeId, symbol: &str, typ: SecurityType) -> Self {
        Self {
            market,
            symbol: symbol.to_string(),
            typ,
        }
    }

    pub fn validate(&self) -> Result<(), String> {
        if self.symbol.is_empty() {
            return Err("security code symbol cannot be empty".into());
        }
        Ok(())
    }
}

impl fmt::Display for SecurityCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}{}", self.market, self.symbol)
    }
}


// Minimal Rust implementation of market flag helper mirroring C++ `GetMarketFlag`
pub fn get_market_flag(market: u8) -> &'static str {
    match market {
        MARKET_SHENZHEN => MARKET_FLAG_SZ,
        MARKET_BEIJING => MARKET_FLAG_BJ,
        MARKET_HONGKONG => MARKET_FLAG_HK,
        MARKET_USA => MARKET_FLAG_US,
        MARKET_SHANGHAI => MARKET_FLAG_SH,
        _ => MARKET_FLAG_SH,
    }
}

/// Build a security code string like C++ `exchange::GetMarketFlag(...) + code`
pub fn security_code(market: u8, code: &str) -> String {
    format!("{}{}", get_market_flag(market), code)
}

/// Detect market id, market flag and pure code from a security string.
/// Mirrors C++ exchange::DetectMarket behavior for common prefixes.
pub fn detect_market(security_code: &str) -> (u8, String, String) {
    let s = security_code.trim();
    let lower = s.to_lowercase();

    // market flags and prefix tables - mirror C++ lists
    // market flag constants
    const MARKET_FLAGS: [&str; 5] = [
        MARKET_FLAG_SH,
        MARKET_FLAG_SZ,
        MARKET_FLAG_BJ,
        MARKET_FLAG_HK,
        MARKET_FLAG_US,
    ];
    let shanghai_main = ["50", "51", "60", "68", "90", "110", "113", "132", "204"];
    let shanghai_special = ["5", "6", "9", "7"];
    let shanghai_other = ["88"];
    let shenzhen_main = [
        "00", "12", "13", "18", "15", "16", "18", "20", "30", "39", "115", "1318",
    ];
    let beijing_main = ["40", "43", "83", "87", "88", "420", "820", "899", "920"];

    // 1) explicit prefix like sh600000 or sh.600000
    for &flag in &MARKET_FLAGS {
        if lower.starts_with(flag) {
            let pure = if s.len() > 2 && &s[2..3] == "." {
                s[3..].to_string()
            } else {
                s[2..].to_string()
            };
            let market_id = match flag {
                MARKET_FLAG_SH => MARKET_SHANGHAI,
                MARKET_FLAG_SZ => MARKET_SHENZHEN,
                MARKET_FLAG_BJ => MARKET_BEIJING,
                MARKET_FLAG_HK => MARKET_HONGKONG,
                MARKET_FLAG_US => MARKET_USA,
                _ => MARKET_SHANGHAI,
            };
            return (market_id, flag.to_string(), pure);
        }
    }

    // 2) explicit suffix like 600000.sh (we follow C++ behavior and strip last 3 chars)
    for &flag in &MARKET_FLAGS {
        let suffix = format!(".{}", flag);
        if lower.ends_with(&suffix) {
            let len = s.len();
            if len > 3 {
                let market_code = flag.to_string();
                let pure = s[..len - 3].to_string();
                let market_id = match flag {
                    MARKET_FLAG_SH => MARKET_SHANGHAI,
                    MARKET_FLAG_SZ => MARKET_SHENZHEN,
                    MARKET_FLAG_BJ => MARKET_BEIJING,
                    MARKET_FLAG_HK => MARKET_HONGKONG,
                    MARKET_FLAG_US => MARKET_USA,
                    _ => MARKET_SHANGHAI,
                };
                return (market_id, market_code, pure);
            }
        }
    }

    // 3) no explicit marker: use prefix tables
    for &p in &shanghai_main {
        if lower.starts_with(p) {
            return (MARKET_SHANGHAI, MARKET_FLAG_SH.to_string(), s.to_string());
        }
    }
    for &p in &shenzhen_main {
        if lower.starts_with(p) {
            return (MARKET_SHENZHEN, MARKET_FLAG_SZ.to_string(), s.to_string());
        }
    }
    for &p in &shanghai_special {
        if lower.starts_with(p) {
            return (MARKET_SHANGHAI, MARKET_FLAG_SH.to_string(), s.to_string());
        }
    }
    for &p in &shanghai_other {
        if lower.starts_with(p) {
            return (MARKET_SHANGHAI, MARKET_FLAG_SH.to_string(), s.to_string());
        }
    }
    for &p in &beijing_main {
        if lower.starts_with(p) {
            return (MARKET_BEIJING, MARKET_FLAG_BJ.to_string(), s.to_string());
        }
    }

    // default fallback: heuristic same as prior implementation
    if s.starts_with('6') {
        (MARKET_SHANGHAI, MARKET_FLAG_SH.to_string(), s.to_string())
    } else {
        (MARKET_SHENZHEN, MARKET_FLAG_SZ.to_string(), s.to_string())
    }
}

// Market id constants (mirror C++ MarketType enum)
pub const MARKET_SHENZHEN: u8 = 0;
pub const MARKET_SHANGHAI: u8 = 1;
pub const MARKET_BEIJING: u8 = 2;
pub const MARKET_HONGKONG: u8 = 21;
pub const MARKET_USA: u8 = 22;

// Market flag string constants (mirror C++ constants)
pub const MARKET_FLAG_SH: &str = "sh";
pub const MARKET_FLAG_SZ: &str = "sz";
pub const MARKET_FLAG_BJ: &str = "bj";
pub const MARKET_FLAG_HK: &str = "hk";
pub const MARKET_FLAG_US: &str = "us";

/// Return true if the given market id and pure code represent an index.
pub fn assert_index_by_market_and_code(market_id: u8, symbol: &str) -> bool {
    let s = symbol.trim();
    if market_id == MARKET_SHANGHAI
        && (s.starts_with("000") || s.starts_with("880") || s.starts_with("881"))
    {
        return true;
    }
    if market_id == MARKET_SHENZHEN && s.starts_with("399") {
        return true;
    }
    // BeiJing index: 899
    if market_id == MARKET_BEIJING && s.starts_with("899") {
        return true;
    }
    false
}

/// Return true if the full security code represents an index.
pub fn assert_index_by_security_code(security_code: &str) -> bool {
    let (market_id, _, code) = detect_market(security_code);
    assert_index_by_market_and_code(market_id, &code)
}

/// If the provided full security code is a ShangHai block (880/881), normalize it in-place
/// to the canonical form (flag+code) and return true. Otherwise return false.
pub fn assert_block_by_security_code(security_code: &mut String) -> bool {
    let (market_id, flag, code) = detect_market(security_code);
    if market_id != MARKET_SHANGHAI {
        return false;
    }
    if !(code.starts_with("880") || code.starts_with("881")) {
        return false;
    }
    *security_code = format!("{}{}", flag, code);
    true
}

/// Return true if the given market id and pure code represent an ETF (ShangHai 510...)
pub fn assert_etf_by_market_and_code(market_id: u8, symbol: &str) -> bool {
    market_id == MARKET_SHANGHAI && symbol.trim().starts_with("510")
}

/// Return true if the given market id and pure code represent an ordinary stock
pub fn assert_stock_by_market_and_code(market_id: u8, symbol: &str) -> bool {
    let s = symbol.trim();
    if market_id == MARKET_SHANGHAI
        && (s.starts_with("60") || s.starts_with("68") || s.starts_with("510"))
    {
        return true;
    }
    if market_id == MARKET_SHENZHEN && (s.starts_with("00") || s.starts_with("30")) {
        return true;
    }
    if market_id == MARKET_BEIJING
        && (s.starts_with("40")
            || s.starts_with("43")
            || s.starts_with("83")
            || s.starts_with("87")
            || s.starts_with("88")
            || s.starts_with("420")
            || s.starts_with("820")
            || s.starts_with("920"))
    {
        return true;
    }
    false
}

/// Return true if the full security code represents a stock
pub fn assert_stock_by_security_code(security_code: &str) -> bool {
    let (market_id, _, code) = detect_market(security_code);
    assert_stock_by_market_and_code(market_id, &code)
}

/// Normalize a security code string to the canonical flag+code format. Empty input => empty output.
pub fn correct_security_code(symbol: &str) -> String {
    if symbol.is_empty() {
        return String::new();
    }
    let (_mid, flag, code) = detect_market(symbol);
    format!("{}{}", flag, code)
}

/// Classification of a security code
#[derive(Debug, PartialEq, Eq)]
pub enum TargetKind {
    Stock,
    Index,
    Block,
    Etf,
}

/// Determine the kind of the security code (block/index/etf/stock)
pub fn assert_code(security_code: &str) -> TargetKind {
    let (market_id, _flag, code) = detect_market(security_code);
    if market_id == MARKET_SHANGHAI {
        // ShangHai: sector prefixes (880/881) -> Block
        if code.starts_with("880") || code.starts_with("881") {
            return TargetKind::Block;
        }
        // ShangHai: 000... -> Index
        if code.starts_with("000") {
            return TargetKind::Index;
        }
        // ShangHai: codes starting with '5' are ETF (per C++ logic)
        if code.starts_with('5') {
            return TargetKind::Etf;
        }
    }
    if market_id == MARKET_SHENZHEN {
        if code.starts_with("399") {
            return TargetKind::Index;
        }
        if code.starts_with("159") {
            return TargetKind::Etf;
        }
    }
    TargetKind::Stock
}

/// Check whether the security code is either an index or a stock (mirrors C++ checkIndexAndStock)
pub fn check_index_and_stock(security_code: &str) -> bool {
    if assert_index_by_security_code(security_code) {
        return true;
    }
    if assert_stock_by_security_code(security_code) {
        return true;
    }
    false
}

#[cfg(test)]
mod assert_tests {
    use super::*;

    #[test]
    fn test_assert_index_and_stock_variants() {
        // Index examples
        assert!(assert_index_by_security_code("sh000001"));
        assert!(assert_index_by_security_code("000001.sh"));
        assert!(assert_index_by_security_code("399001"));

        // Stock examples
        assert!(assert_stock_by_security_code("sh600000"));
        assert!(assert_stock_by_security_code("sz000001"));

        // ETF example
        assert!(assert_etf_by_market_and_code(MARKET_SHANGHAI, "510500"));
        assert_eq!(assert_code("sh880001"), TargetKind::Block);
        assert_eq!(assert_code("sh000001"), TargetKind::Index);
        assert_eq!(assert_code("sz399001"), TargetKind::Index);
        assert_eq!(assert_code("sh600000"), TargetKind::Stock);
    }

    #[test]
    fn test_assert_block_and_correct_code() {
        let mut s = String::from("880001");
        // Without flag, detect_market will infer ShangHai and assert_block should normalize
        assert!(assert_block_by_security_code(&mut s));
        assert_eq!(s, "sh880001");

        let mut s2 = String::from("sz000001");
        assert!(!assert_block_by_security_code(&mut s2));

        assert_eq!(correct_security_code("600000"), "sh600000".to_string());
        assert_eq!(correct_security_code(""), "".to_string());
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_market_prefix_and_suffix() {
        let cases = vec![
            ("sh600000", 1u8, "sh", "600000"),
            ("sh.600000", 1u8, "sh", "600000"),
            ("600000.sh", 1u8, "sh", "600000"),
            ("sz000001", 0u8, "sz", "000001"),
            ("000001.sz", 0u8, "sz", "000001"),
            ("000001", 0u8, "sz", "000001"),
            ("600000", 1u8, "sh", "600000"),
            ("400001", 2u8, "bj", "400001"),
            ("880000", 1u8, "sh", "880000"),
            ("115000", 0u8, "sz", "115000"),
            ("hk00700", 21u8, "hk", "00700"),
            ("00700.hk", 21u8, "hk", "00700"),
            ("usAAPL", 22u8, "us", "AAPL"),
        ];

        let cases = cases
            .into_iter()
            .map(|(input, exp_id, exp_flag, exp_pure)| {
                let id_const = match exp_id {
                    0u8 => MARKET_SHENZHEN,
                    1u8 => MARKET_SHANGHAI,
                    2u8 => MARKET_BEIJING,
                    21u8 => MARKET_HONGKONG,
                    22u8 => MARKET_USA,
                    other => other,
                };
                (input, id_const, exp_flag, exp_pure)
            })
            .collect::<Vec<_>>();

        for (input, exp_id, exp_flag, exp_pure) in cases {
            let (id, flag, pure) = detect_market(input);
            assert_eq!(id, exp_id, "id mismatch for input: {}", input);
            assert_eq!(
                flag,
                exp_flag.to_string(),
                "flag mismatch for input: {}",
                input
            );
            assert_eq!(
                pure.to_lowercase(),
                exp_pure.to_lowercase(),
                "pure mismatch for input: {} -> {}",
                input,
                pure
            );
        }
    }
}

File: quant1x/exchange/code.py
from enum import Enum
from typing import Tuple, List

# Constants
MARKET_SHANGHAI = "sh"
MARKET_SHENZHEN = "sz"
MARKET_BEIJING = "bj"
MARKET_HONGKONG = "hk"
MARKET_USA = "us"

MARKET_FLAGS = ["sh", "sz", "SH", "SZ", "bj", "BJ", "hk", "HK", "us", "US"]

SHANGHAI_MAIN_BOARD_PREFIXES = ["50", "51", "60", "68", "90", "110", "113", "132", "204"]
SHANGHAI_SPECIAL_PREFIXES = ["5", "6", "9", "7"]
SECTOR_PREFIXES = ["880", "881"]
SHENZHEN_MAIN_BOARD_PREFIXES = ["00", "12", "13", "18", "15", "16", "18", "20", "30", "39", "115", "1318"]
BEIJING_MAIN_BOARD_PREFIXES = ["40", "43", "83", "87", "88", "420", "820", "899", "920"]

class MarketType(Enum):
    ShenZhen = 0
    ShangHai = 1
    BeiJing = 2
    HongKong = 21
    USA = 22

class TargetKind(Enum):
    STOCK = 0
    INDEX = 1
    BLOCK = 2
    ETF = 3

def starts_with(s: str, prefixes: List[str]) -> bool:
    """
    检查字符串是否以给定的任意前缀开头
    
    Args:
        s (str): 要检查的目标字符串
        prefixes (List[str]): 可能的前缀列表
    
    Returns:
        bool: 如果字符串以任一前缀开头则返回True，否则返回False
    """
    return any(s.startswith(prefix) for prefix in prefixes)

def ends_with(s: str, suffixes: List[str]) -> bool:
    """
    检查字符串是否以给定的任一后缀结尾
    
    Args:
        s (str): 要检查的字符串
        suffixes (List[str]): 可能的后缀列表
    
    Returns:
        bool: 如果字符串以任一后缀结尾则返回True，否则返回False
    """
    return any(s.endswith(suffix) for suffix in suffixes)

def get_security_code(market: MarketType, symbol: str) -> str:
    """
    根据市场类型和证券代码生成完整的证券代码字符串
    
    Args:
        market (MarketType): 市场类型枚举值
        symbol (str): 原始证券代码
    
    Returns:
        str: 根据市场规则生成的完整证券代码字符串
    """
    if market == MarketType.USA:
        return MARKET_USA + symbol
    elif market == MarketType.HongKong:
        return MARKET_HONGKONG + symbol[:5]
    elif market == MarketType.BeiJing:
        return MARKET_BEIJING + symbol[:6]
    elif market == MarketType.ShenZhen:
        return MARKET_SHENZHEN + symbol[:6]
    else:
        return MARKET_SHANGHAI + symbol[:6]

def get_market(symbol: str) -> str:
    """
    根据股票代码判断其所属的市场
    
    Args:
        symbol (str): 股票代码字符串，可能包含市场前缀或后缀标识
        
    Returns:
        str: 市场标识符的小写形式，如'sh'表示上海市场，'sz'表示深圳市场，'bj'表示北京市场，'hk'表示香港市场，'us'表示美国市场
    
    说明:
        1. 首先检查代码开头或结尾是否包含市场标识符
        2. 如果没有显式市场标识，则根据代码前缀判断所属市场
        3. 默认返回上海市场标识
    """
    code = symbol.strip()
    market = MARKET_SHANGHAI

    if starts_with(code, MARKET_FLAGS):
        market = code[:2].lower()
    elif ends_with(code, MARKET_FLAGS):
        market = code[-2:].lower()
    elif starts_with(code, SHANGHAI_MAIN_BOARD_PREFIXES):
        market = MARKET_SHANGHAI
    elif starts_with(code, SHENZHEN_MAIN_BOARD_PREFIXES):
        market = MARKET_SHENZHEN
    elif starts_with(code, SHANGHAI_SPECIAL_PREFIXES):
        market = MARKET_SHANGHAI
    elif starts_with(code, SECTOR_PREFIXES):
        market = MARKET_SHANGHAI
    elif starts_with(code, BEIJING_MAIN_BOARD_PREFIXES):
        market = MARKET_BEIJING
    elif code.isalpha():
        market = MARKET_USA
    
    return market

def get_market_id(symbol: str) -> MarketType:
    """
    根据股票代码符号获取对应的市场类型
    
    Args:
        symbol (str): 股票代码符号，例如 '600000' 或 '000001'
    
    Returns:
        MarketType: 返回对应的市场枚举值，包括:
            - MarketType.ShangHai: 上海市场
            - MarketType.ShenZhen: 深圳市场 
            - MarketType.BeiJing: 北京市场
            - MarketType.HongKong: 香港市场
            - MarketType.USA: 美国市场
            默认返回上海市场
    """
    market = get_market(symbol)
    if market == MARKET_SHANGHAI:
        return MarketType.ShangHai
    if market == MARKET_SHENZHEN:
        return MarketType.ShenZhen
    if market == MARKET_BEIJING:
        return MarketType.BeiJing
    if market == MARKET_HONGKONG:
        return MarketType.HongKong
    if market == MARKET_USA:
        return MarketType.USA
    return MarketType.ShangHai

def get_market_flag(market_id: MarketType) -> str:
    """
    根据市场类型ID获取对应的市场标志字符串
    
    Args:
        market_id (MarketType): 市场类型枚举值
    
    Returns:
        str: 对应市场的标志字符串，可能的返回值包括:
            - MARKET_SHENZHEN: 深圳市场
            - MARKET_BEIJING: 北京市场
            - MARKET_HONGKONG: 香港市场
            - MARKET_USA: 美国市场
            - MARKET_SHANGHAI: 上海市场(默认值)
    """
    if market_id == MarketType.ShenZhen:
        return MARKET_SHENZHEN
    elif market_id == MarketType.BeiJing:
        return MARKET_BEIJING
    elif market_id == MarketType.HongKong:
        return MARKET_HONGKONG
    elif market_id == MarketType.USA:
        return MARKET_USA
    else:
        return MARKET_SHANGHAI

def detect_market(symbol: str) -> Tuple[MarketType, str, str]:
    """
    根据股票代码识别所属市场类型并提取纯净代码
    
    Args:
        symbol (str): 原始股票代码字符串，可能包含市场标识前缀/后缀(如'sh', 'sz')或特殊前缀
    
    Returns:
        Tuple[MarketType, str, str]: 返回三元组，包含:
            - 市场类型枚举值(MarketType)
            - 市场代码字符串(如'sh', 'sz')
            - 去除市场标识后的纯净股票代码
    
    Note:
        支持以下识别方式:
        1. 市场标识前缀(如'sh600000')
        2. 市场标识后缀(如'600000.sh')
        3. 上海/深圳/北京主板特殊前缀
        4. 板块指数特殊前缀
    """
    pure_code = symbol.strip()
    market_code = MARKET_SHANGHAI

    if starts_with(pure_code, MARKET_FLAGS):
        market_code = pure_code[:2].lower()
        if len(pure_code) > 2 and pure_code[2] == '.':
            pure_code = pure_code[3:]
        else:
            pure_code = pure_code[2:]
    elif ends_with(pure_code, MARKET_FLAGS):
        market_code = pure_code[-2:].lower()
        pure_code = pure_code[:-3]
    elif starts_with(pure_code, SHANGHAI_MAIN_BOARD_PREFIXES):
        market_code = MARKET_SHANGHAI
    elif starts_with(pure_code, SHENZHEN_MAIN_BOARD_PREFIXES):
        market_code = MARKET_SHENZHEN
    elif starts_with(pure_code, SHANGHAI_SPECIAL_PREFIXES):
        market_code = MARKET_SHANGHAI
    elif starts_with(pure_code, SECTOR_PREFIXES):
        market_code = MARKET_SHANGHAI
    elif starts_with(pure_code, BEIJING_MAIN_BOARD_PREFIXES):
        market_code = MARKET_BEIJING

    market_id = MarketType.ShangHai
    if market_code == MARKET_SHANGHAI:
        market_id = MarketType.ShangHai
    elif market_code == MARKET_SHENZHEN:
        market_id = MarketType.ShenZhen
    elif market_code == MARKET_BEIJING:
        market_id = MarketType.BeiJing
    elif market_code == MARKET_HONGKONG:
        market_id = MarketType.HongKong
    elif market_code == MARKET_USA:
        market_id = MarketType.USA
    
    return market_id, market_code, pure_code

def assert_index_by_market_and_code(market_id: MarketType, symbol: str) -> bool:
    """
    根据市场类型和证券代码验证是否为指数证券
    
    Args:
        market_id (MarketType): 市场类型枚举值（ShangHai/ShenZhen/BeiJing）
        symbol (str): 证券代码字符串
    
    Returns:
        bool: 如果代码符合该市场的索引规则返回True，否则返回False
    """
    if market_id == MarketType.ShangHai and starts_with(symbol, ["000", "880", "881"]):
        return True
    if market_id == MarketType.ShenZhen and starts_with(symbol, ["399"]):
        return True
    if market_id == MarketType.BeiJing and starts_with(symbol, ["899"]):
        return True
    return False

def assert_index_by_security_code(security_code: str) -> bool:
    """
    根据证券代码验证是否为指数证券
    
    Args:
        security_code (str): 证券代码字符串
    
    Returns:
        bool: 如果是指数证券返回True，否则返回False
    """
    market_id, _, code = detect_market(security_code)
    return assert_index_by_market_and_code(market_id, code)

def assert_block_by_security_code(security_code: str) -> Tuple[bool, str]:
    """
    根据证券代码验证是否为板块代码，并返回处理后的代码
    
    Args:
        security_code (str): 待验证的证券代码
    
    Returns:
        Tuple[bool, str]: 返回元组，第一个元素表示是否为板块代码，
                         第二个元素为处理后的代码(原代码或带前缀的代码)
    """
    market_id, flag, code = detect_market(security_code)
    if market_id != MarketType.ShangHai or not starts_with(code, SECTOR_PREFIXES):
        return False, security_code
    return True, flag + code

def assert_etf_by_market_and_code(market_id: MarketType, symbol: str) -> bool:
    """
    检查给定市场和代码是否匹配ETF的规则
    
    Args:
        market_id (MarketType): 市场类型枚举值
        symbol (str): 证券代码
    
    Returns:
        bool: 如果市场是上海市场且代码以'510'开头则返回True，否则返回False
    """
    return market_id == MarketType.ShangHai and starts_with(symbol, ["510"])

def assert_stock_by_market_and_code(market_id: MarketType, symbol: str) -> bool:
    """
    根据市场类型和股票代码验证股票是否属于指定市场
    
    Args:
        market_id (MarketType): 市场类型枚举值(ShangHai/ShenZhen/BeiJing)
        symbol (str): 股票代码字符串
    
    Returns:
        bool: 如果股票代码符合该市场的编码规则返回True，否则返回False
    """
    if market_id == MarketType.ShangHai and starts_with(symbol, ["60", "68", "510"]):
        return True
    if market_id == MarketType.ShenZhen and starts_with(symbol, ["00", "30"]):
        return True
    if market_id == MarketType.BeiJing and starts_with(symbol, ["40", "43", "83", "87", "88", "420", "820", "920"]):
        return True
    return False

def assert_stock_by_security_code(security_code: str) -> bool:
    """
    根据证券代码验证股票代码格式是否正确
    
    Args:
        security_code (str): 证券代码字符串，格式应符合市场检测要求
    
    Returns:
        bool: 如果该证券代码符合股票代码格式规则则返回True，否则返回False
    """
    market_id, _, code = detect_market(security_code)
    return assert_stock_by_market_and_code(market_id, code)

def correct_security_code(symbol: str) -> str:
    """
    根据给定的证券代码符号，修正并返回标准化的证券代码。
    
    Args:
        symbol (str): 输入的证券代码符号，可以是任意格式的证券代码表示
    
    Returns:
        str: 修正后的标准化证券代码，格式为"市场标志+证券代码"
        
    Note:
        如果输入为空字符串，将返回空字符串
    """
    if not symbol:
        return ""
    _, m_flag, m_symbol = detect_market(symbol)
    return m_flag + m_symbol

def assert_code(security_code: str) -> TargetKind:
    market_id, _, code = detect_market(security_code)
    if market_id == MarketType.ShangHai:
        if starts_with(code, SECTOR_PREFIXES):
            return TargetKind.BLOCK
        if starts_with(code, ["000"]):
            return TargetKind.INDEX
        if starts_with(code, ["5"]):
            return TargetKind.ETF
    if market_id == MarketType.ShenZhen:
        if starts_with(code, ["399"]):
            return TargetKind.INDEX
        if starts_with(code, ["159"]):
            return TargetKind.ETF
    if market_id == MarketType.BeiJing and starts_with(code, ["899"]):
        return TargetKind.INDEX
    return TargetKind.STOCK

def check_index_and_stock(security_code: str) -> bool:
    if assert_index_by_security_code(security_code):
        return True
    if assert_stock_by_security_code(security_code):
        return True
    return False
